### PR TITLE
Added SQS Publisher for Notifications

### PR DESCRIPTION
**Pull request by massenz**
This PR completes the ability for SM to receive `Events` and publish `Notifications` to SQS.
Test coverage mostly exceeds 70% for all packages, excluding the `api` package, due to the fact that it is mostly made up of Protobuf generated classes (and also we are not testing all the marshal/unmarshal JSON code).

```
api      coverage: 27.2%
pubsub   coverage: 81.0%
server   coverage: 73.5%
storage  coverage: 81.6%
```

This needs to be tested using two containers running, Redis and Localstack (see the `Dockercompose.yaml` spec); use `make test` to replicate.

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -66,7 +66,7 @@ func SetLogLevel(level log.LogLevel) {
 // that path params are testable.
 func NewRouter() *mux.Router {
```
Comment (mklumpp1) [inline server/server.go:67]: is server going to be exposed as lambda under a gateway?

**Review by mklumpp1 — APPROVED** [review]

File: Makefile [hunk 2]
```diff
@@ -2,36 +2,39 @@
...
+api/statemachine.pb.go: protos/statemachine.proto
 	protoc --proto_path=protos/ \
                --go_out=api/ \
                --go_opt=paths=source_relative \
                protos/*.proto
 
-protos: api/statemachine.pb.go
-
-all: protos cmd/server.go
+all: api/statemachine.pb.go cmd/server.go
```
[the remark this answers is not in the document]
Comment (massenz) [reply Makefile:13]: container in EKS.
I already have (elsewhere) the skeleton of a `Dockerfile` for a Go binary, it's just a matter of writing up the `make container` one-liner

one thing to discuss (separately) is how to deploy (and connect) to Redis.

File: api/fsm.go [hunk 3]
```diff
@@ -136,10 +136,10 @@ func (x *Configuration) HasState(state string) bool {
     return false
 }
 
-// CheckValid checks that the Configuration is valid and that the current FSM's `state` is one of
+// CheckValid checks that the Configuration is valid and that the current FSM `state` is one of
 // the allowed states in the Configuration.
 //
-// We also check that the reported FSM's ConfigId, matches the Configuration's name, version.
+// We also check that the reported FSM ConfigId, matches the Configuration's name, version.
 func (x *ConfiguredStateMachine) CheckValid() bool {
```
[the remark this answers is not in the document]
Comment (massenz) [reply api/fsm.go:143]: ah! so, this is what Google Protobuf `protoc` for Go uses - if you use something else (eg, `cfsm` or whatever, you get a warning that you're changing the var name) so I'm using `x` here too for consistency.

But totally agree!

File: api/fsm.go [hunk 4]
```diff
@@ -159,7 +159,7 @@ func (x *Configuration) CheckValid() error {
         return MissingStatesConfigurationError
     }
     if x.StartingState == "" || !x.HasState(x.StartingState) {
```
[the remark this answers is not in the document]
Comment (massenz) [reply api/fsm.go:161]: not only that, but actually, this needs splitting into two separate checks (done) so that the error message reflects the error of an empty state
thanks

File: cmd/server.go [hunk 5]
```diff
@@ -31,35 +31,37 @@ import (
...
+    var dlqTopic = flag.String("errors", "",
+        "The name of the Dead-Letter Queue ("+"DLQ) in SQS to post errors to; if not "+
+            "specified, the DLQ will not be used")
     flag.Parse()
 
     logger := log.NewLog("statemachine")
-    logger.Level = log.INFO
 
     var host = "0.0.0.0"
     if *localOnly {
```
[the remark this answers is not in the document]
Comment (massenz) [reply cmd/server.go:67]: the variable is (possibly) poorly named - what this controls is whether the server listens only on `localhost` or on all NICs for incoming requests

the only real difference between "local only" and AWS is around the URIs for Redis/SQS and those are externalized.

File: cmd/server.go [hunk 6]
```diff
@@ -80,33 +82,50 @@ func main() {
...
-        logger.Panic("support for Kafka not implemented")
-    } else if *sqsTopic != "" {
-        logger.Info("Connecting to SQS Topic: %s", *sqsTopic)
-        sub = pubsub.NewSqsSubscriber(eventsCh)
-    } else {
-        logger.Warn("No event broker configured, state machines will not be able to receive events")
+    var errorsCh chan pubsub.EventErrorMessage = nil
+
+    if *eventsTopic == "" {
+        logger.Fatal(fmt.Errorf("no event topic configured, state machines will not " +
```
[the remark this answers is not in the document]
Comment (massenz) [reply cmd/server.go:97]: `logging.Fatal` takes an `error` (not a `string`) so `fmt` is needed to create one (and there is no `Error()` method there - without the `f`)

`Fatalf` emits a panic, which I'm not really a fan of (it's really hard to parse and figure out what happened)
`logging.Fatal` emits a more civilized log line and then exits (via `os.Exit()`)

File: pubsub/listener.go [hunk 7]
```diff
@@ -19,33 +19,12 @@
...
-type ListenerOptions struct {
-    EventsChannel        <-chan EventMessage
-    NotificationsChannel chan<- EventErrorMessage
-    StatemachinesStore   storage.StoreManager
-    ListenersPoolSize    int8
-}
-
 func NewEventsListener(options *ListenerOptions) *EventsListener {
     return &EventsListener{
         logger:        logging.NewLog("Listener"),
```
[the remark this answers is not in the document]
Comment (massenz) [reply pubsub/listener.go:30]: shall we say that Go's `logging` is... basic and not particularly easy to configure (there is no concept of log level, and it's really hard to configure the format of the emitted log) - so I've created my own `logging` module, where we have something closer to Slf4j (at least with several log levels, a `shouldLog` etc.)

Alas, there is no Lombok's `@Slf4j` in Go (or any way to make it easy to automatically create a logger at creation) - what you see here is the closest one can get (also worth noting that Go's structs don't really have a "constructor" - again, what you see here is as close as it gets)

That `logging` module you see here will be soon refactored in its own repo and externalized - it's got nothing to do with SM

As for CloudWatch - well, yes, of course, these logs go to stdout and from there, one presumes, to CW, don't they?

File: pubsub/listener.go [hunk 8]
```diff
@@ -60,25 +39,11 @@ func (l *EventsListener) SetLogLevel(level logging.LogLevel) {
...
-        fmt.Fprintf(&msgBuf, " (%s)", detail)
-    }
-    l.logger.Error(msgBuf.String())
-
-    var errorMsg = EventErrorMessage{
-        Error:       *NewEventProcessingError(err),
-        ErrorDetail: detail,
-        Message:     &msg,
-    }
+func (l *EventsListener) PostErrorNotification(error *EventErrorMessage) {
```
[the remark this answers is not in the document]
Comment (massenz) [reply pubsub/listener.go:42]: not to mention confusing with the number `1`
although this seems to be a bit of a Go "idiom" to use a one-letter "receiver"

changed to `listener` everywhere, though, I agree with you.

File: pubsub/listener.go [hunk 9]
```diff
@@ -87,20 +52,20 @@ func (l *EventsListener) ListenForMessages() {
     for event := range l.events {
         l.logger.Debug("Received event %s", event)
         if event.Destination == "" {
-            l.PostErrorNotification(event, fmt.Errorf("no destination for event"), "")
+            l.PostErrorNotification(ErrorMessage(fmt.Errorf("no destination for event"), &event))
```
[the remark this answers is not in the document]
Comment (massenz) [reply pubsub/listener.go:55]: as above, I need an `error` type there, not a `string`

File: pubsub/sqs_sub.go [hunk 10]
```diff
@@ -38,11 +42,27 @@ type SqsSubscriber struct {
 
 // NewSqsSubscriber will create a new `Subscriber` to listen to
 // incoming api.Event from a SQS `queue`.
-func NewSqsSubscriber(eventsChannel chan<- EventMessage) *SqsSubscriber {
-    // TODO: need to confirm this works when running inside an EKS Node.
-    sess := session.Must(session.NewSessionWithOptions(session.Options{
-        SharedConfigState: session.SharedConfigEnable,
-    }))
+func NewSqsSubscriber(eventsChannel chan<- EventMessage, sqsUrl *string) *SqsSubscriber {
```
[the remark this answers is not in the document]
Comment (massenz) [reply pubsub/sqs_sub.go:45]: this is so embarrassing... :relaxed: 
:+1:

[the remark this answers is not in the document]
Comment (massenz) [reply pubsub/sqs_sub.go:45]: BTW - yes, under the hood they are Singletons (and if the Go client is anything like Python's - not even thread-safe) but this is taken care of by the AWS machinery, I can just re-create them as many times as I want (I still get the same object - I think)

File: pubsub/sqs_sub.go [hunk 11]
```diff
@@ -38,11 +42,27 @@ type SqsSubscriber struct {
...
+            SharedConfigState: session.SharedConfigEnable,
+            Config: aws.Config{
+                Endpoint: sqsUrl,
+                Region:   &region,
+            },
+        }))
+    }
     client := sqs.New(sess)
     return &SqsSubscriber{
         logger:          logging.NewLog("SQS-Sub"),
```
[the remark this answers is not in the document]
Comment (massenz) [reply pubsub/sqs_sub.go:68]: so, please note there *is* a difference between initializing it here (it's a different logger "per class" to the extent that one considers a Go struct resembling a Java class) while in `init()` you can only initialize package-wide vars
of course, one could argue that there only needs to be one logger per package (maybe?) but this is not the intent here

also, we are not "passing" them around, we are "instantiating" them inside a "constructor" and keeping a pointer reference to the instantiated object (so there's very little, if any, cost in copying the SqS*** structs)

[the remark this answers is not in the document]
Comment (massenz) [reply pubsub/sqs_sub.go:68]: oh, and both Publisher/Subscriber (and a few other types) all live in the `pubsub` package

File: server/server.go (hunk 1 above)
Comment (massenz) [reply server/server.go:67]: nope - it's a "real" server, running as an EKS container
(although, now you mention it, one could also conceive it the other way round, as a collection of Lambdas - a couple behind the G/W, and the other two connected to SQS, while Redis acts as the repository of state)

not how I've designed it so far, though (which is not to say it was The Right Way :laughing: )

**Comment by massenz** [comment]
Thanks for review @mklumpp1 :bow:
